### PR TITLE
Add max acceleration mm per s2, feedrates and jerk save and restor du…

### PR DESCRIPTION
**Pull request by 3d-gussner**
…ring power panic

Thanks to @wavexx in #2456

### CODE REVIEW

**Comment by gudnimg** [comment]
Using  `eeprom_update_block` instead of four 4x `eeprom_update_float` will save a lot of flash. It’s not faster though, I wonder how much time all of these EEPROM writes add to the power panic. 

Quick look at the datasheet for ATmega2560 gives typical EEPROM write time 3.3ms. 60 bytes is ~200ms… assuming its 3.3ms per byte 🤔

**Comment by 3d-gussner** [comment]
>It’s not faster though, I wonder how much time all of these EEPROM writes add to the power panic.
Quick look at the datasheet for ATmega2560 gives typical EEPROM write time 3.3ms. 60 bytes is ~200ms… assuming its 3.3ms per byte 🤔

Tested on MK404
- Operating printer with normal PrusaSlicer profile
  - PowerPanic reports `UVLO - end 164` which is ~164ms (on MK404 not real printer)
- Factory reset printer with normal PrusaSlicer profile 
  - PowerPanic reports `UVLO - end 165` which is ~165ms (on MK404 not real printer)

Also these values are stored with `M500` in eeprom, so as long the print profiles don't change a lot I expect not a big increase of time needed while power panic.

Also checked the PrusaSlicer-settings and the Machine limits values didn't change as far I can tell.

Another test with multiple changes in the gcode setting different  jerk, accel and feedrates increased to `UVLO - end 180` (on MK404 not real printer)

If we want to save these values we need to deal with the time delay when these settings are different to previous eeprom values.

The eeprom_update still costs some `ms` as it has to read the eeprom and verify if it changed.


@gudnimg Thanks for the hint with `eeprom_update_block_notify` saved 800 bytes of flash.

**Comment by gudnimg** [comment]
> Also these values are stored with M500 in eeprom, so as long the print profiles don't change a lot I expect not a big increase of time needed while power panic.

Thats a good point. I forgot yesterday that the update functions only do EEPROM write if the value has changed 👍

File: Firmware/eeprom.h [hunk 1]
```diff
@@ -385,6 +385,35 @@ static_assert(sizeof(Sheets) == EEPROM_SHEETS_SIZEOF, "Sizeof(Sheets) is not EEP
 | 0x0C7F 3199 | bool    | EEPROM_UVLO_Z_LIFTED                  | 00h 0        | 00h                   | Power Panic Z axis NOT lifted                     | Power Panic  | D3 Ax0c7f C1
 | ^           | ^       | ^                                     | 01h 1        | 01h                   | Power Panic Z axis lifted                         | ^            | ^
 | 0x0C7d 3197 | uint16  | EEPROM_UVLO_EXTRUDE_MINTEMP           | 0-305        | afh 175               | Power Panic Extrude mintemp                       | Power Panic  | D3 Ax0c7d C2
+| 0x0C6D 3181 | float   | EEPROM_UVLO_ACCELL_MM_S2_NORMAL       | ???          | ff ff ff ffh          | Power Panic acceleration mm per s2 normal         | Power Panic  | D3 Ax0c6d C16
```
Comment (gudnimg) [inline Firmware/eeprom.h]: Change `float` to `uint32`

File: Firmware/eeprom.h [hunk 2]
```diff
@@ -385,6 +385,35 @@ static_assert(sizeof(Sheets) == EEPROM_SHEETS_SIZEOF, "Sizeof(Sheets) is not EEP
...
+| ^           | ^       | ^                                     | ???          | ^                     | E-axis                                            | ^            | D3 Ax0c79 C4
+| ^           | ^       | ^                                     | ???          | ^                     | Z-axis                                            | ^            | D3 Ax0c75 C4
+| ^           | ^       | ^                                     | ???          | ^                     | Y-axis                                            | ^            | D3 Ax0c71 C4
+| ^           | ^       | ^                                     | ???          | ^                     | X-axis                                            | ^            | D3 Ax0c6d C4
+| 0x0C5D 3165 | float   | EEPROM_UVLO_ACCELL_MM_S2_SILENT       | ???          | ff ff ff ffh          | Power Panic acceleration mm per s2 silent         | Power Panic  | D3 Ax0c5d C16
```
Comment (gudnimg) [inline Firmware/eeprom.h]: Change `float` to `uint32`

File: Firmware/eeprom.h [hunk 3]
```diff
@@ -626,9 +655,16 @@ static Sheets * const EEPROM_Sheets_base = (Sheets*)(EEPROM_SHEETS_BASE);
 #define EEPROM_CUSTOM_MENDEL_NAME (EEPROM_FILENAME_EXTENSION-17) //char[17]
 #define EEPROM_UVLO_Z_LIFTED (EEPROM_CUSTOM_MENDEL_NAME-1) //bool
 #define EEPROM_UVLO_EXTRUDE_MINTEMP (EEPROM_UVLO_Z_LIFTED-2) //uint16_t
-
+#define EEPROM_UVLO_ACCELL_MM_S2_NORMAL (EEPROM_UVLO_EXTRUDE_MINTEMP-4*4) // 4 x float
+#define EEPROM_UVLO_ACCELL_MM_S2_SILENT (EEPROM_UVLO_ACCELL_MM_S2_NORMAL-4*4) // 4 x float
```
Comment (gudnimg) [inline Firmware/eeprom.h]: Change `float` to `uint32`

File: Firmware/eeprom.h [hunk 4]
```diff
@@ -626,9 +655,16 @@ static Sheets * const EEPROM_Sheets_base = (Sheets*)(EEPROM_SHEETS_BASE);
...
+#define EEPROM_UVLO_MAX_FEEDRATE_NORMAL (EEPROM_UVLO_ACCELL_MM_S2_SILENT-4*4) // 4 x float
+#define EEPROM_UVLO_MAX_FEEDRATE_SILENT (EEPROM_UVLO_MAX_FEEDRATE_NORMAL-4*4) // 4 x float
+#define EEPROM_UVLO_MIN_FEEDRATE (EEPROM_UVLO_MAX_FEEDRATE_SILENT-4) //float
+#define EEPROM_UVLO_MIN_TRAVEL_FEEDRATE (EEPROM_UVLO_MIN_FEEDRATE-4) //float
+#define EEPROM_UVLO_MIN_SEGMENT_TIME_US (EEPROM_UVLO_MIN_TRAVEL_FEEDRATE-4) //unit32_t
```
Comment (gudnimg) [inline Firmware/eeprom.h]: typo, unit32_t --> uint32_t

File: Firmware/power_panic.cpp [hunk 5]
```diff
@@ -191,13 +191,26 @@ void uvlo_() {
...
+    // Save min extrude temp
+    eeprom_update_word_notify((uint16_t*)EEPROM_UVLO_EXTRUDE_MINTEMP, extrude_min_temp);
+#endif //PREVENT_DANGEROUS_EXTRUDE
+    // Save max acceleration mm per s2
+    eeprom_update_block_notify(cs.max_acceleration_mm_per_s2_normal, (uint32_t *)EEPROM_UVLO_ACCELL_MM_S2_NORMAL, sizeof(cs.max_acceleration_mm_per_s2_normal));
+    eeprom_update_block_notify(cs.max_acceleration_mm_per_s2_silent, (uint32_t *)EEPROM_UVLO_ACCELL_MM_S2_SILENT, sizeof(cs.max_acceleration_mm_per_s2_silent));
+    // Save feedrates
+    eeprom_update_block_notify(cs.max_feedrate_normal, (float *)EEPROM_UVLO_MAX_FEEDRATE_NORMAL, sizeof(cs.max_feedrate_normal));
+    eeprom_update_block_notify(cs.max_feedrate_silent, (float *)EEPROM_UVLO_MAX_FEEDRATE_SILENT, sizeof(cs.max_feedrate_silent));
+    // Save Jerk
```
Comment (gudnimg) [inline Firmware/power_panic.cpp]: Just my opinion, but I feel these comments aren't useful:

```
// Save min extrude temp
// Save max acceleration mm per s2
// Save feedrates
// Save Jerk
```

I would remove them as the variable names themselves tell me the same information

File: Firmware/power_panic.cpp [hunk 6]
```diff
@@ -426,8 +439,20 @@ bool recover_machine_state_after_power_panic() {
...
+    eeprom_read_block(cs.max_acceleration_mm_per_s2_normal, (uint32_t *)EEPROM_UVLO_ACCELL_MM_S2_NORMAL, sizeof(cs.max_acceleration_mm_per_s2_normal));
+    eeprom_read_block(cs.max_acceleration_mm_per_s2_silent, (uint32_t *)EEPROM_UVLO_ACCELL_MM_S2_SILENT, sizeof(cs.max_acceleration_mm_per_s2_silent));
+    // Recover feedrates
+    eeprom_read_block(cs.max_feedrate_normal, (float *)EEPROM_UVLO_MAX_FEEDRATE_NORMAL, sizeof(cs.max_feedrate_normal));
+    eeprom_read_block(cs.max_feedrate_silent, (float *)EEPROM_UVLO_MAX_FEEDRATE_SILENT, sizeof(cs.max_feedrate_silent));
+    // Recover jerk
+    cs.minimumfeedrate = eeprom_read_float((float *)EEPROM_UVLO_MIN_FEEDRATE);
+    cs.mintravelfeedrate = eeprom_read_float((float *)EEPROM_UVLO_MIN_TRAVEL_FEEDRATE);
+    cs.min_segment_time_us = eeprom_read_dword((uint32_t *)EEPROM_UVLO_MIN_SEGMENT_TIME_US);
+    eeprom_read_block(cs.max_jerk, (float *)EEPROM_UVLO_MAX_JERK, sizeof(cs.max_jerk));
```
Comment (gudnimg) [inline Firmware/power_panic.cpp]: Similar to another comment of mine, are these comments really useful?

```
// Recover min extrude temp
// Recover max acceleration mm per s2
// Recover feedrates
// Recover jerk
```

**Review by gudnimg — CHANGES_REQUESTED** [review]
PR looks good to me, just some minor typos that need fixing

**Comment by 3d-gussner** [comment]
> PR looks good to me, just some minor typos that need fixing

@gudnimg Thanks for the review, fixed the typos

**Review by gudnimg — APPROVED** [review]

**Pull request merged**